crates/wasm-encoder/src/component/types.rs

```diff
@@ -0,0 +1,574 @@
+                bytes.extend(encoders::u32(*index));
+                bytes.push(OUTER_ALIAS_TYPE);
+            }
+        }
+    }
+}
+
+/// Represents an alias.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Alias<'a> {
+    /// The alias is to an instance export.
+    InstanceExport {
+        /// The index of the instance being aliased.
+        instance: u32,
+        /// The name of the export.
+        name: &'a str,
+        /// The kind of the definition being aliased.
+        kind: DefinitionKind,
+    },
+    /// The alias is to an index in an outer module.
+    Outer {
+        /// The count of the outer module.
+        ///
+        /// The count starts at 0 which denotes the current module.
+        count: u32,
+        /// The index from the outer module's index space being aliased.
+        index: OuterAliasIndex,
+    },
+}
+
+impl Alias<'_> {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::InstanceExport {
+                instance,
+                name,
+                kind,
+            } => {
+                bytes.push(ALIAS_TYPE_INSTANCE_EXPORT);
+                bytes.extend(encoders::u32(*instance));
+                bytes.extend(encoders::str(name));
+                kind.encode(bytes);
+            }
+            Self::Outer { count, index } => {
+                bytes.push(ALIAS_TYPE_OUTER);
+                bytes.extend(encoders::u32(*count));
+                index.encode(bytes);
+            }
+        }
+    }
+}
+
+/// Represents a named interface type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct NamedInterfaceType<'a> {
+    name: &'a str,
+    ty: InterfaceType,
+}
+
+impl NamedInterfaceType<'_> {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        bytes.extend(encoders::str(self.name));
+        self.ty.encode(bytes);
+    }
+}
+
+/// Represents results of an adapter function.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum AdapterFunctionResults<'a> {
+    /// The adapter function returns zero results.
+    None,
+    /// The adapter function returns one unnamed result.
+    Single(InterfaceType),
+    /// The function returns multiple named results.
+    ///
+    /// The length of the slice must be greater than one.
+    Multiple(&'a [NamedInterfaceType<'a>]),
+}
+
+impl AdapterFunctionResults<'_> {
+    pub(crate) fn encode(&self, bytes: &mut Vec<u8>) {
+        match self {
+            Self::None => bytes.push(0x00),
+            Self::Single(ty) => {
+                bytes.push(0x01);
+                ty.encode(bytes);
+            }
+            Self::Multiple(types) => {
+                assert!(types.len() > 1);
+                bytes.extend(encoders::u32(types.len() as u32));
+                for ty in *types {
+                    ty.encode(bytes);
+                }
+            }
+        }
+    }
+}
+
+/// Represents a type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Type<'a> {
```

Comment:
Given the recursive and index-bound nature of a `Type` as well as the number of lifetimes involved here, this might be more deserving of an encode-style interface where each variant here, if necessary, has a builder which internally builds up the encoding but can be incrementally added to. That would mean that you don't have to have all the `InstanceTypeDef` items resident at the same time, nor do there need to be vectors-in-vectors and such in case the most end up working around the lifetimes.

Reply:
I agree, this API is a bit hard to use.

I'll replace with a builder pattern.

Reply:
I've replaced this enum (and all the lifetime-involved types) with encoders for the various types; there are nested vectors when defining instance and module types because they need to know how many definitions are being defined _before_ their encoding, but the singular builders borrow their parent vector to encode the types.

I'm open to a different approach to this too.